presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/UnwrapDateFunctionInPredicate.java

```diff
@@ -0,0 +1,249 @@
+        System.out.println("Literal: " + literal + " (" + literal.getClass().getName() + ")");
+        System.out.println("Literal Type: " + literal.getType());
+
+        // Ensure the function has exactly one argument
+        if (function.getArguments().size() != 1) {
+            System.out.println("Function has unexpected number of arguments.");
+            return Optional.empty();
+        }
+
+        RowExpression column = function.getArguments().get(0);
+
+        // Unwrap CAST if present on the column side
+        if (column instanceof CallExpression) {
+            CallExpression castExpression = (CallExpression) column;
+            if (castExpression.getDisplayName().equalsIgnoreCase("CAST") && castExpression.getArguments().size() == 1) {
+                column = castExpression.getArguments().get(0);
+            }
+        }
+
+        // Check that the column is a valid reference
+        if (!(column instanceof VariableReferenceExpression || column instanceof InputReferenceExpression)) {
+            return Optional.empty();
+        }
+
+        // If the literal is of type VARCHAR (io.airlift.slice.Slice), try to parse it as a date string.
+        if (literal.getType().getJavaType().equals(Slice.class)) {
+            try {
+                Slice slice = (Slice) literal.getValue();
+                String dateString = slice.toStringUtf8();
+                long epochDay = parseEpochDay(dateString);
+                literal = new ConstantExpression(epochDay, DateType.DATE);
+                System.out.println("Converted VARCHAR literal to DateType using epochDay: " + epochDay);
+            } catch (Exception e) {
+                System.out.println("Failed to parse VARCHAR as DateType: " + e.getMessage());
+                return Optional.empty();
+            }
+        }
+
+        // At this point the literal must be a DateType
+        if (!(literal.getType() instanceof DateType)) {
+            System.out.println("Literal type is still not DateType, exiting.");
+            return Optional.empty();
+        }
+
+        // Extract the epoch day from the literal (as a long)
+        Object value = literal.getValue();
+        long dateDays;
+        if (value instanceof Number) {
+            dateDays = ((Number) value).longValue();
+        }
+        else {
+            System.out.println("Unexpected literal value type: " + value.getClass());
+            return Optional.empty();
+        }
+
+        // Presto TIMESTAMP literals are stored as microseconds since the Unix epoch.
+        final long MICROSECONDS_PER_DAY = 86400000000L;
+        long lowerBoundMicros = dateDays * MICROSECONDS_PER_DAY;
+        long upperBoundMicros = (dateDays + 1) * MICROSECONDS_PER_DAY;
+
+        ConstantExpression lowerTimestampLiteral = new ConstantExpression(lowerBoundMicros, TimestampType.TIMESTAMP);
+        ConstantExpression upperTimestampLiteral = new ConstantExpression(upperBoundMicros, TimestampType.TIMESTAMP);
```

Comment:
You could also just write a `CallExpression` which is a `CAST` from date to timestamp. Then, either wait for the ExpressionOptimizer to constant-fold your value into another `ConstantExpression`, or use the `RowExpressionInterpreter` to turn it into a timestamp directly here. I think I told Maks something similar on his PR. https://github.com/Zumaks/presto/pull/2#discussion_r2008306839